### ISSUE TITLE
libcxx setting ignored during test_package

### ISSUE THREAD

**Issue by cinghiale** [conan-io/conan#403]
Hi all,

I'm trying to package this lib `https://github.com/cpp-netlib/uri` for conan.

This is my first package, so it's very probable that I'm missing something obvious, but I have a link error (related with the libstdc++ abi) when using conan test_package

The conanfile.py of the package:

```python
class UriConan(ConanFile):
    name = "uri"
    version = "master"
    revision = "60c0da339ae387fa54373eeca312c7f10f526120"
    license = "BSD"
    url = "Boost Software License - Version 1.0 - August 17th, 2003"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = "shared=False", "fPIC=True"

    def config(self):
        self.settings.compiler.libcxx = "libstdc++11"

    def source(self):
        self.run("git clone https://github.com/cpp-netlib/uri.git")
        self.run("cd uri && git checkout %s" % self.revision)
        self.run("cd uri && git submodule init")
        self.run("cd uri && git submodule update")

    def build(self):
        shared = "-DBUILD_SHARED_LIBS=ON" if self.options.shared else ""
        fpic = "-DCMAKE_POSITION_INDEPENDENT_CODE=TRUE" if self.options.fPIC else ""
        cmake = CMake(self.settings)
        self.run("cd uri && cmake . %s %s %s" % (cmake.command_line, shared, fpic))
        self.run("cd uri && cmake --build . %s" % cmake.build_config)

    def package(self):
        #import pdb; pdb.set_trace()
        self.copy("*.hpp", dst="include/network", src="uri/include/network")
        self.copy("libnetwork-uri.a", dst="lib", src="uri/src", keep_path=False)
        self.copy("libnetwork-uri.so*", dst="lib", src="uri/src", keep_path=False)
        self.copy("*.lib", dst="lib", src="uri/src", keep_path=False)
        self.copy("*.dll", dst="bin", src="uri/src", keep_path=False)

    def package_info(self):
        self.cpp_info.libs = ["network-uri"]
```

and the conanfile.py of the test_package

```python
channel = os.getenv("CONAN_CHANNEL", "testing")
username = os.getenv("CONAN_USERNAME", "dvd")


class UriTestConan(ConanFile):
        settings = "os", "compiler", "build_type", "arch"
        requires = "uri/master@%s/%s" % (username, channel)
        generators = "cmake"

        def config(self):
                self.settings.compiler.libcxx = "libstdc++11"

        def build(self):
                cmake = CMake(self.settings)
                self.run('cmake "%s" %s' % (self.conanfile_directory, cmake.command_line))
                self.run("cmake --build . %s" % cmake.build_config)

        def imports(self):
                self.copy("*.dll", "bin", "bin")
                self.copy("*.dylib", "bin", "bin")

        def test(self):
                os.chdir("bin")
                self.run(".%sexample" % os.sep)
```

The package forcibly sets the "compiler.libcxx" (this is just for test, I should check the compiler and set libcxx only for gcc, or probably better use the settings dict to disallow the usage of `libstdc++`)
as well as the `test project` conanfile.

But it looks like that the project settings are ignored, I've edited conans/client/cmake.py to print the flags and I obtain

    ['-DCONAN_COMPILER="gcc"', '-DCONAN_COMPILER_VERSION="5.4"', '-DCONAN_LIBCXX="libstdc++"']

It looks like that during the test_package the system wide setting (the one in ~/.conan/conan.conf) takes the precedence

**Comment by memsharded** [comment]
Hi!

First, a clarification about settings and options. The settings are something that the package creator can not define. They are defined by the package consumer. You cannot specify that your settings are os="Windows" or compiler="Visual Studio", because the developer trying to use your package can be using gcc under ubuntu. The most a package creator can do about settings is to restrict them, for example, throwing and Exception in ``config()`` for certain settings, so the package consumer gets warned that such code will not work for they platform.

The same applies for the libstdc++11, you cannot force it, it depends on the package consumers compiler and system. Thats is why the settings have no defaults in the package recipe, as opposed as the options. You can have a package option for choosing between shared or static library, and choose a default for it.

When you are consuming a package (or developing it as a package creator), you can specify the settings in two places:
- In the command line: ``-s compiler.libcxx=libstdc++11``
- In the conf file ``~/.conan/conan.conf``` you have the defaults that are used when you do not specify explicitly settings in the command line. You can change them if you want too.


Another tip: as you are packaging "master" version, and coding a specific commit, from conan version 0.11, you can have options with the value ``ANY``, which you could default to the commit you want, and override the commit if you want to try with a different one.

Please have a look to the libstdc++11 issue, I will try to have a look to it when possible (I am out of office, nearly no internet connection), in which OS are you working? Ubuntu 16.04? Also, please copy & paste the few lines with linking errors of your output. Thanks!

**Comment by cinghiale** [comment]
So I should change the config() method of the package in something like this:
```python
    def config(self):
        if self.settings.compiler == "gcc" and self.settings.compiler.libcxx != "libstdc++11":
            log message
            raise exception
```

Any advice about which exception raise?

IIUC the config() of the test_package is correct, it can force the usage of the new abi; I'm using ubuntu 16.04,  this is the tail of the output of `conan test_package`

```
PROJECT: Generated cmake created conanbuildinfo.cmake
PROJECT: Generated conaninfo.txt
['-DCONAN_COMPILER="gcc"', '-DCONAN_COMPILER_VERSION="5.4"', '-DCONAN_LIBCXX="libstdc++"']
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan C++ stdlib: libstdc++
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dvd/Projects/personal/tmp/conan-uri/test_package/build/da39a3ee5e6b4b0d3255bfef95601890afd80709
Scanning dependencies of target example
[ 50%] Building CXX object CMakeFiles/example.dir/example.cpp.o
[100%] Linking CXX executable bin/example
CMakeFiles/example.dir/example.cpp.o: In function `network::uri::uri<char [20]>(char const (&) [20])':
example.cpp:(.text._ZN7network3uriC2IA20_cEERKT_[_ZN7network3uriC5IA20_cEERKT_]+0x6c): undefined reference to `network::uri::initialize(std::string const&)'
collect2: error: ld returned 1 exit status
CMakeFiles/example.dir/build.make:94: recipe for target 'bin/example' failed
make[2]: *** [bin/example] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/example.dir/all' failed
make[1]: *** [CMakeFiles/example.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
ERROR: Error 512 while executing cmake --build . 
```

The third line is the flags passed to the cmake generator

**Comment by memsharded** [comment]
Yes, you can do that, you can raise a normal ``Exception("your message")`` it will be translated to a ConanException and output to the user as an Error (in red).

Just a note, only do that if you are totally sure that the code for such package will NOT link against the libstdc++. Your code can be C++11 and link against the libstdc++, that shouldn't be a problem.

Yes, you are right, it is not that the ``config()`` cannot set the libcxx value, is that it shouldn't. If I try to use your package and I am in a computer without libstdc++11, I wouldn't be able, even if the package would compile and link against libstdc++11, because the value is now hardwired, so by definition it is not a setting anymore, it would be a constant :)

I will try to have a look to the project when I have access to my development environment, thanks very much for your detailed feedback.

**Comment by memsharded** [comment]
Hi @cinghiale ,

I have full access now to my dev environment :)

I have checked, and I cannot find a repo for this package recipe. Do you have them in a github repo, so I can clone it and test it locally, and collaborate? Thanks very much!

**Comment by cinghiale** [comment]
Hi @memsharded I started working again on that recipe just yesterday. Give me a couple of days and I come back with an url

**Comment by cinghiale** [comment]
I put the recipe here: https://gitlab.com/cinghiale/conan-cppnet/tree/master
It is for a different, but related, package.

The conanfile of the test_package (https://gitlab.com/cinghiale/conan-cppnet/blob/master/test_package/conanfile.py) sets the libcxx flag to `libstdc++11` but the package itself is builded with the settings taken from `~/.conan/conan.conf`.

IIUC the test_package should be treated like a normal program and thus it should be able to dictate the libcxx

**Comment by memsharded** [comment]
Hi! I have been able to clone your repo and test it, it worked in Ubuntu 16, gcc 5.4, good job! A few comments:

- The test inside test_package, doesn't seem to use cppnet, just boost. Is this correct? Something I am missing? Such test is intended to use cppnet code, so I would expect including it.

- Try to avoid raw 2.7 python ``print``, as conan also works in python 3 and there are a few users that use it. Using it, will not allow them to use your package. Try to remember to use ``print()`` function instead

- Try to use ``configure()`` instead of ``config()``. The latest is being deprecated.

No, you shouldn't set ``libcxx`` in your ``test_package`` either. It is not the test package that defines the existence of a ``libstdc++11`` library, it is the system, your computer setup. In my case, I am running gcc 5.4 in Ubuntu 16, and that setup has a ``libstdc++11``. So the input to conan as a setting (that can be defined from ``conan.conf`` default or in the command line) is that I have ``libstdc++11``, together with Linux, gcc, 5.4 and x86_64 settings.

Consider it, exactly the same as the other settings, does it make sense to have this configure()?

```python
    def config(self):
        self.settings.os = "Linux"
        self.settings.compiler = "gcc"
        self.settings.compiler = "5.4"
        self.settings.compiler.libcxx = "libstdc++11"
```

It doesn't make sense. It restricts the usage (calling ``conan test_package``) for any other configuration. But **cppnetlib** might work with other compilers as clang, other compiler versions as 4.9-5.3, etc. There is no reason why it won't built against a ``libstdc++`` library.

Even if the system allows to link some modules to ``libstdc++`` and others to ``libstdc++11``, I'd say it is not a good idea. When trying to link everything together it is very likely that problems will occur. In this case, it could happen that the cppnet library is actually not being used yet by the example program in test_package, so those problems doesn't occur. The standard library version used should be the same for the whole program.

Please try  using cppnet in the example, then if there are any problem, removing the ``config()`` method from test_package and running, and tell me. Thanks!

**Comment by cinghiale** [comment]
cppnet lives inside the boost include/namespace, see: http://cpp-netlib.org/0.12.0/index.html#hello-world so i think the test is ok.

What are the differences between `configure()` and `config()`?

**Comment by memsharded** [comment]
``config()`` is executed two times, one before and one after propagating option values to the package recipe. That leads to undesired side effects if there is configuration with settings that affect options and options that affect settings, and tricky solutions like using counters were required.

It has been deprecated (recipes with it keep working) in favor of ``config_options()`` executed before propagating options values, and ``configure()`` which is executed after, and is the one needed for most configurations. The ``config_options()`` is only required for configuring options depending on the settings, for example, if the ``shared`` option can be only True in Linux, but not in windows, it is possible to do that in ``config_options()`` so if a Windows user try to build such a package with shared=True, it will complain. Most other configurations can be done in ``configure()``.

I didn't realize that cppnet is inside boost namespace. Then it is ok, sure. Thanks for telling.

**Comment by cinghiale** [comment]
Hi @memsharded according to your definition your recipe: [conan-fmt](https://github.com/memsharded/conan-fmt/blob/testing/conanfile.py) should be updated to move:

```python
if self.settings.os == "Windows":
     self.options.remove("fPIC")
```

in the `config_options()` method?

**Comment by memsharded** [comment]
Yes, exactly, good catch. I have just submitted a fix (https://github.com/memsharded/conan-fmt/blob/testing/conanfile.py), it would be like:

```python
def config_options(self):
        if self.settings.os == "Windows":
            self.options.remove("fPIC")

def configure(self):
        if self.options.header_only:
            self.settings.clear()
            self.options.remove("shared")
            self.options.remove("fPIC")
        else:
            self.build_policy = None
```

Not only that, but also removing unused options for header_only. That cannot be done in ``config_options``, as the value for ``header_only`` is still not assigned. The advantage of ``config_options`` is that in windows if I try to ``conan install -o fmt:fPIC=True`` it will error with a message saying that fPIC is not an option for me.

I have pushed the changes, but some CI job is failing I have to have a look. Thanks for pointing this out!
